llvm/test/CodeGen/AMDGPU/udiv.ll

```diff
@@ -2619,37 +2619,38 @@ define i64 @v_test_udiv64_mulhi_fold(i64 %arg) {
 ; VI-LABEL: v_test_udiv64_mulhi_fold:
 ; VI:       ; %bb.0:
 ; VI-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
-; VI-NEXT:    v_mov_b32_e32 v4, 0xa7c5
-; VI-NEXT:    v_mul_u32_u24_e32 v3, 0x500, v4
-; VI-NEXT:    v_mul_hi_u32_u24_e32 v2, 0x500, v4
-; VI-NEXT:    v_add_u32_e32 v3, vcc, 0x4237, v3
-; VI-NEXT:    v_addc_u32_e32 v5, vcc, 0, v2, vcc
-; VI-NEXT:    v_add_u32_e32 v6, vcc, 0xa9000000, v3
-; VI-NEXT:    s_mov_b32 s6, 0xfffe7960
-; VI-NEXT:    v_mad_u64_u32 v[2:3], s[4:5], v6, s6, 0
-; VI-NEXT:    v_addc_u32_e32 v7, vcc, v5, v4, vcc
-; VI-NEXT:    v_mul_lo_u32 v4, v7, s6
+; VI-NEXT:    s_mov_b32 s4, 0x346d900
+; VI-NEXT:    s_add_u32 s4, 0x4237, s4
```

Comment:
Do you have a plan to fold this add?

Reply:
Not yet, it looks like this:
```
      t420: v2i32 = BUILD_VECTOR Constant:i32<16951>, Constant:i32<0>
    t421: i64 = bitcast t420
      t414: v2i32 = BUILD_VECTOR Constant:i32<54974720>, Constant:i32<0>
    t415: i64 = bitcast t414
  t190: i64 = add t421, t415
```
We can't do a generic combine for the bitcast I think. I tried it and it caused loops. I think the easiest would be to make a DAG equivalent of the GISel helper that looks through splats and just returns the value directly.

Reply:
We might be able to replace `dyn_cast<ConstantSDNode>` checks in SelectionDAG::FoldConstantArithmetic with ComputeKnownBits calls that looks for isConstant scalar values?

Reply:
I don't understand what the issue is here. Why can't we fold (bitcast (build_vector constant:i32, constant:i32)) -> constant:i64 ?

Reply:
We already have a `ConstantFoldBITCASTofBUILD_VECTOR` fold but it only runs before LegalizeTypes/Ops. Some targets need the bitcasts apparently. I tried doing it more often but only in specific cases (<=64 bit ints) but I still get infinite loops everywhere.

I think we need to add some helper to get splat values instead of checking ConstantSDNode when constant folding add. What do you think?

Reply:
Shouldn't it just be a matter of legality checks to avoid the loops?